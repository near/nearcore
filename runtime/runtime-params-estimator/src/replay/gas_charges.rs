use super::Visitor;
use std::collections::BTreeMap;
use std::io::Write;

/// Visitor that tracks which DB operations are charged with gas and which are
/// for free.
#[derive(Default)]
pub(super) struct ChargedVsFree {
    in_fn_call: bool,
    storage_op_indent: usize,

    free_gets: u64,
    free_gets_size: u64,
    charged_gets: u64,
    charged_gets_size: u64,
}

impl Visitor for ChargedVsFree {
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        _key: &[u8],
        _col: &str,
    ) -> anyhow::Result<()> {
        self.eval_label(out, indent, op, &BTreeMap::new())?;
        if op != "GET" {
            return Ok(());
        }
        if self.in_fn_call {
            self.charged_gets += 1;
            self.charged_gets_size += size.unwrap_or(0);
        } else {
            self.free_gets += 1;
            self.free_gets_size += size.unwrap_or(0);
        }
        Ok(())
    }

    fn eval_storage_op(
        &mut self,
        _out: &mut dyn Write,
        indent: usize,
        _op: &str,
        _dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        self.storage_op_indent = indent + 2;
        self.in_fn_call = true;
        Ok(())
    }

    fn eval_label(
        &mut self,
        _out: &mut dyn Write,
        indent: usize,
        _label: &str,
        _dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        if indent < self.storage_op_indent {
            self.in_fn_call = false;
        }
        Ok(())
    }

    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        writeln!(
            out,
            "{:>8} free gets with total size of    {:>8}",
            self.free_gets, self.free_gets_size
        )?;
        writeln!(
            out,
            "{:>8} charged gets with total size of {:>8}",
            self.charged_gets, self.charged_gets_size
        )?;
        writeln!(
            out,
            "{:>7.2}% of gets uncharged",
            100.0 * self.free_gets as f64 / (self.charged_gets + self.free_gets) as f64
        )?;
        writeln!(
            out,
            "{:>7.2}% of gets total size uncharged",
            100.0 * self.free_gets_size as f64
                / (self.charged_gets_size + self.free_gets_size) as f64
        )?;

        self.free_gets = 0;
        self.free_gets_size = 0;
        self.charged_gets = 0;
        self.charged_gets_size = 0;

        Ok(())
    }
}
