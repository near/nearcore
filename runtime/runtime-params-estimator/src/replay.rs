use anyhow::Context;
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{self, Write};
use std::path::PathBuf;
use std::str::SplitWhitespace;
use tracing::log::error;

use self::fold_db_ops::FoldDbOps;
use self::gas_charges::ChargedVsFree;

mod cache_stats;
mod fold_db_ops;
mod gas_charges;

#[derive(clap::Parser)]
pub(crate) struct ReplayCmd {
    trace: PathBuf,
    #[clap(subcommand)]
    mode: ReplayMode,
    /// Only show data for a specific smart contract, specified by account id.
    #[clap(long)]
    account: Option<String>,
}

#[derive(Clone, Copy, clap::Subcommand, Debug)]
pub(crate) enum ReplayMode {
    /// Print DB accesses and cache statistics for the entire trace.
    CacheStats,
    /// Print DB accesses per receipt.
    ReceiptDbStats,
    /// Print DB accesses and cache statistics per receipt.
    ReceiptCacheStats,
    /// Print DB accesses per chunk.
    ChunkDbStats,
    /// Print DB accesses and cache statistics per chunk.
    ChunkCacheStats,
    /// Go over DB operations and print how much of it is paid for with gas.
    GasCharges,
}

impl ReplayCmd {
    pub(crate) fn run(&self, out: &mut dyn Write) -> anyhow::Result<()> {
        let file = File::open(&self.trace)?;
        self.run_on_input(io::BufReader::new(file), out)
    }

    fn run_on_input(&self, input: impl io::BufRead, out: &mut dyn Write) -> anyhow::Result<()> {
        let mut visitor = self.build_visitor();
        for line in input.lines() {
            let line = line?;
            if let Err(e) = visitor.eval_line(out, &line) {
                error!("ERROR: {e} for input line: {line}");
            }
        }
        visitor.flush(out)?;
        Ok(())
    }

    fn build_visitor(&self) -> Box<dyn Visitor> {
        match &self.mode {
            ReplayMode::CacheStats => {
                Box::new(FoldDbOps::new().with_cache_stats().account_filter(self.account.clone()))
            }
            ReplayMode::ChunkDbStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks())
            }
            ReplayMode::ChunkCacheStats => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with per-chunk statistics");
                }
                Box::new(FoldDbOps::new().chunks().with_cache_stats())
            }
            ReplayMode::ReceiptDbStats => {
                Box::new(FoldDbOps::new().receipts().account_filter(self.account.clone()))
            }
            ReplayMode::ReceiptCacheStats => Box::new(
                FoldDbOps::new().receipts().with_cache_stats().account_filter(self.account.clone()),
            ),
            ReplayMode::GasCharges => {
                if self.account.is_some() {
                    unimplemented!("account filter does not work with gas charges");
                }
                Box::new(ChargedVsFree::default())
            }
        }
    }
}

/// Interface for processing an IO trace line-by-line.
///
/// The visitor default methods take over the basic parsing of the IO trace.
/// Specific implementations can extract the data of interest and process it
/// further as necessary, without duplicating the parsing code.
trait Visitor {
    /// The root entry point of the visitors.
    ///
    /// This function takes a raw input line as input without any preprocessing.
    /// A visitor may choose to overwrite this function for full control but the
    /// intention is that the default implementation takes over the basic
    /// parsing and visitor implementations define their behaviour using the
    /// other trait methods.
    fn eval_line(&mut self, out: &mut dyn Write, line: &str) -> anyhow::Result<()> {
        if let Some(indent) = line.chars().position(|c| !c.is_whitespace()) {
            let mut tokens = line.split_whitespace();
            if let Some(keyword) = tokens.next() {
                match keyword {
                    "GET" | "SET" | "UPDATE_RC" => {
                        let col = tokens.next().context("missing column field in DB operation")?;
                        let mut key_str = tokens.next().context("missing key in DB operation")?;
                        if key_str.starts_with('"') {
                            key_str = &key_str[1..key_str.len() - 1];
                        }
                        let key = near_fmt::Bytes::from_str(key_str).map_err(|e| {
                            anyhow::anyhow!("failed to decode key {key_str} because {e}")
                        })?;
                        let dict = extract_key_values(tokens)?;
                        let size: Option<u64> = dict.get("size").map(|s| s.parse()).transpose()?;
                        self.eval_db_op(out, indent, keyword, size, &key, col)?;
                    }
                    "storage_read" | "storage_write" | "storage_remove" | "storage_has_key" => {
                        let op = tokens.next();
                        if op.is_none() {
                            return Ok(());
                        }

                        let dict = extract_key_values(tokens)?;
                        self.eval_storage_op(out, indent, keyword, &dict)?;
                    }
                    other_label => {
                        let dict = extract_key_values(tokens)?;
                        self.eval_label(out, indent, other_label, &dict)?;
                    }
                }
            }
        }
        Ok(())
    }

    /// Gets called for every trie storage operation.
    ///
    /// A storage operation is a layer above DB operations, one storage
    /// operation can cause many DB operations, or it could be resolved from
    /// cache without any DB operations.
    fn eval_storage_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, op, dict);
        Ok(())
    }

    /// Gets called for every DB operation.
    fn eval_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
        col: &str,
    ) -> anyhow::Result<()> {
        if col == "State" {
            self.eval_state_db_op(out, indent, op, size, key)
        } else {
            Ok(())
        }
    }

    /// Gets called for every DB operation on the state column.
    fn eval_state_db_op(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        op: &str,
        size: Option<u64>,
        key: &[u8],
    ) -> anyhow::Result<()> {
        let (_, _, _, _, _) = (out, indent, op, size, key);
        Ok(())
    }

    /// Opening spans that are not storage or DB operations.
    fn eval_label(
        &mut self,
        out: &mut dyn Write,
        indent: usize,
        label: &str,
        dict: &BTreeMap<&str, &str>,
    ) -> anyhow::Result<()> {
        let (_, _, _, _) = (out, indent, label, dict);
        Ok(())
    }

    /// Print accumulated internal state.
    ///
    /// This function is called at the end of a replayed trace. Visitor
    /// implementation usually choose to call this more often. For example, once
    /// for each receipt.
    fn flush(&mut self, out: &mut dyn Write) -> anyhow::Result<()> {
        let _ = out;
        Ok(())
    }
}

fn extract_key_values<'a>(
    mut tokens: SplitWhitespace<'a>,
) -> anyhow::Result<BTreeMap<&'a str, &'a str>> {
    let mut dict = BTreeMap::new();
    while let Some(key_val) = tokens.next() {
        let (key, value) =
            key_val.split_once('=').context("key-value pair delimited by `=` expected")?;
        dict.insert(key, value);
    }
    Ok(dict)
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::{ReplayCmd, ReplayMode};

    /// These inputs are real mainnet traffic for the given block heights.
    /// Each trace contains two chunks in one shard.
    /// In combination, they cover an interesting variety of content.
    ///  Shard 0: Many receipts, but no function calls.
    ///  Shard 1: Empty chunks without txs / receipts.
    ///  Shard 2: Some receipts and txs, but no function calls.
    ///  Shard 3: Many function calls with lots of storage accesses.
    const INPUT_TRACES: &[&str] = &[
        "75220100-75220101.s0.io_trace",
        "75220100-75220101.s1.io_trace",
        "75220100-75220101.s2.io_trace",
        "75220100-75220101.s3.io_trace",
    ];

    /// A synthetic trace that should be a bit easier to read and verify.
    const SYNTHETIC_TRACE: &str = r#"
GET BlockHeader "`fAkeHeAd3R`" size=6000
GET BlockInfo "`FAk31nf0`" size=1
apply_transactions shard_id=0
    process_state_update 
        apply num_transactions=1 shard_cache_hit=10 shard_cache_miss=1
            process_transaction tx_hash=txHash0 shard_cache_miss=1 shard_cache_hit=20
                GET State "'stateKey0'" size=300
            process_receipt receipt_id=id0 predecessor=system receiver=alice.near id=id0 shard_cache_miss=2 shard_cache_hit=6
                GET State "'stateKey1'" size=10
                GET State "'stateKey2'" size=20
            process_receipt receipt_id=id1 predecessor=system receiver=bob.near id=id1 shard_cache_hit=18 shard_cache_miss=3 shard_cache_too_large=1
                GET State "'stateKey3'" size=100
                GET State "'stateKey4'" size=200
                GET State "'stateKey5'" size=9000
            process_receipt receipt_id=id2 predecessor=system receiver=alice.near id=id2 shard_cache_miss=1 shard_cache_hit=6
                GET State "'stateKey6'" size=30
                attached_deposit 
                input 
                register_len 
                read_register 
                storage_read READ key=StorageKey0 size=1000 tn_db_reads=20 tn_mem_reads=0 shard_cache_hit=19 shard_cache_miss=1
                    GET State "'stateKey7'" size=5
            process_receipt receipt_id=id3 predecessor=utiha.near receiver=alice.near id=id3 shard_cache_miss=0 shard_cache_hit=15
            GET State "'stateKey8'" size=300
        GET State "'stateKey9'" size=400
GET State "'stateKey10'" size=500
"#;

    #[test]
    fn test_cache_stats() {
        check_replay_mode(ReplayMode::CacheStats);
    }

    #[test]
    fn test_chunk_cache_stats() {
        check_replay_mode(ReplayMode::ChunkCacheStats);
    }

    #[test]
    fn test_chunk_db_stats() {
        check_replay_mode(ReplayMode::ChunkDbStats);
    }

    #[test]
    fn test_gas_charges() {
        check_replay_mode(ReplayMode::GasCharges);
    }

    #[test]
    fn test_receipt_cache_stats() {
        check_replay_mode(ReplayMode::ReceiptCacheStats);
    }

    #[test]
    fn test_receipt_db_stats() {
        check_replay_mode(ReplayMode::ReceiptDbStats);
    }

    #[track_caller]
    fn check_replay_mode(mode: ReplayMode) {
        for trace_name in INPUT_TRACES {
            let dir = env!("CARGO_MANIFEST_DIR");
            let trace_path = std::path::Path::new(dir).join("res").join(trace_name);
            let cmd = ReplayCmd { trace: trace_path, mode, account: None };
            let mut buffer = Vec::new();
            cmd.run(&mut buffer).unwrap_or_else(|e| {
                panic!("command should not fail for input {trace_name}, failure was {e}")
            });
            let output = String::from_utf8(buffer).unwrap_or_else(|e| {
                panic!("invalid output for input {trace_name}, failure was {e}")
            });
            insta::assert_snapshot!(format!("{mode:?}-{trace_name}"), output);
        }
    }

    #[test]
    fn test_account_filter_cache_stats() {
        check_account_filter(ReplayMode::CacheStats);
    }

    #[test]
    fn test_account_filter_receipt_cache_stats() {
        check_account_filter(ReplayMode::ReceiptCacheStats);
    }

    #[test]
    fn test_account_filter_receipt_db_stats() {
        check_account_filter(ReplayMode::ReceiptDbStats);
    }

    #[track_caller]
    fn check_account_filter(mode: ReplayMode) {
        let account = Some("alice.near".to_owned());
        // trace path not used, will be read from in-memory input instead
        let trace = PathBuf::new();
        let cmd = ReplayCmd { trace, mode, account };
        let mut buffer = Vec::new();
        cmd.run_on_input(SYNTHETIC_TRACE.as_bytes(), &mut buffer).expect("failed replaying");
        let output =
            String::from_utf8(buffer).unwrap_or_else(|e| panic!("invalid output, failure was {e}"));
        insta::assert_snapshot!(format!("account_filter_{mode:?}"), output);
    }
}
