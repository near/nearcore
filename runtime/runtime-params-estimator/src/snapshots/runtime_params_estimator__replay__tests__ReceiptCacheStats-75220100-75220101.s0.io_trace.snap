---
source: runtime/runtime-params-estimator/src/replay.rs
expression: output
---
      process_transaction tx_hash=6wbch2CCuvqhbnknr7HhNW9PoyK6FSUmoCGJKpKEahvi block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   37 State  
        DB GET               37 requests for a total of     4473 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           2 /  37  (shard-cache/DB)
        SHARD CACHE          5.13% hit rate
        CHUNK CACHE not accessed

      process_transaction tx_hash=8P2uMiA96mwRDnoqzwQijEHc1eFpZNJYJFKxHz69sYgj block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   30 State  
        DB GET               30 requests for a total of     5549 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           6 /  30  (shard-cache/DB)
        SHARD CACHE         16.67% hit rate
        CHUNK CACHE not accessed

      process_transaction tx_hash=GREvhhJebp73HsJh8dXJYgdMURBPbpMebDti4AVi6UdC block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   3 State  
        DB GET                3 requests for a total of      165 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          18 /   3  (shard-cache/DB)
        SHARD CACHE         85.71% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=app.nearcrowd.near receipt_id=HSkLwmSscYqj6DekmxRTUACkBUAUS7sUVBnxPC2wKVj8 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   64 State  
        DB GET               64 requests for a total of   549798 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          3 requests for a total of      288 B
        STORAGE WRITE         3 requests for a total of      288 B
        TRIE NODES (guest)  146 /   2 /  50  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           2 /  14  (shard-cache/DB)
        SHARD CACHE          5.88% hit rate,   5.97% if removing 1 too large nodes from total
        CHUNK CACHE         73.74% hit rate

      process_receipt receiver=app.nearcrowd.near receipt_id=HRR3EXitfuoA7yFoa1239GW13ZErmdq6J8vt1MoZgpiu block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   28 State  
        DB GET               28 requests for a total of   545285 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          3 requests for a total of      288 B
        STORAGE WRITE         3 requests for a total of      288 B
        TRIE NODES (guest)  138 /   0 /  27  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          13 /   1  (shard-cache/DB)
        SHARD CACHE         31.71% hit rate,  32.50% if removing 1 too large nodes from total
        CHUNK CACHE         83.64% hit rate

      process_receipt receiver=app.nearcrowd.near receipt_id=28ozHpuL1WZ3ExMm3UrNQYco31XHHaBiUXbyMD1firye block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   17 State  
        DB GET               17 requests for a total of   543852 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          3 requests for a total of      224 B
        STORAGE WRITE         4 requests for a total of      305 B
        TRIE NODES (guest)  116 /   0 /  16  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          13 /   1  (shard-cache/DB)
        SHARD CACHE         43.33% hit rate,  44.83% if removing 1 too large nodes from total
        CHUNK CACHE         87.88% hit rate

      process_receipt receiver=6f25217bc63c86d36e0cdc620d05b2f1381eb43eac560d4f1261380b255e2176 receipt_id=Fp2sJPeBSPX5eXNanJ5Jh1shgBKn6omvUCWcUNEAMd4W block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   26 State  
        DB GET               26 requests for a total of     5023 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           4 /  26  (shard-cache/DB)
        SHARD CACHE         13.33% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=76248d053a8e8b47cfecab60c5f6f1befe74bcbd923c4ef5950bfeb2945fadff receipt_id=6d7ZrM9yztyWvAVtYbA4nQhtfchiyXbbBmJzYyMe3nht block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   24 State  
        DB GET               24 requests for a total of     4519 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           4 /  24  (shard-cache/DB)
        SHARD CACHE         14.29% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=7a5de36efdbcd23489dcfd7a3c63966b9bcbddfbe4ba8db7d1df109ecd37ef5e receipt_id=wDN56cf2bf3sz16RzF6H3UaC1pGnYJ7XFvUdzLTiTS3 block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   28 State  
        DB GET               28 requests for a total of     4505 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           6 /  28  (shard-cache/DB)
        SHARD CACHE         17.65% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=27b8a4173e15829829159b39c620e9a5da908a3f68d94a70cd489188d478f37f receipt_id=D3noeb84wCn3hym3TyGarXfJZJvnFEBiQRk1fQ3W1jws block=FhsQS1zgTLpZjYitRKevStmzJfUgweUTjQA1rUeNW6QR
        GET   28 State  
        DB GET               28 requests for a total of     5009 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           4 /  28  (shard-cache/DB)
        SHARD CACHE         12.50% hit rate
        CHUNK CACHE not accessed

      process_transaction tx_hash=2Aw4SPFoD6ia1Vr6NxqHpxKmK73JM9ZBw2HisUzgaUB5 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   30 State  
        DB GET               30 requests for a total of     5607 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           4 /  30  (shard-cache/DB)
        SHARD CACHE         11.76% hit rate
        CHUNK CACHE not accessed

      process_transaction tx_hash=EZ9GhJxWQgcTpEQBH2JutymJU8jcUAzGE45a4wym3mK9 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   26 State  
        DB GET               26 requests for a total of     4545 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          10 /  26  (shard-cache/DB)
        SHARD CACHE         27.78% hit rate
        CHUNK CACHE not accessed

      process_transaction tx_hash=6ysp5YTw2enoSzjn5hZBp6K8NunF7noEHesY65PF4WDv block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   22 State  
        DB GET               22 requests for a total of     4177 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          10 /  22  (shard-cache/DB)
        SHARD CACHE         31.25% hit rate
        CHUNK CACHE not accessed

      process_transaction tx_hash=8GExB7nr4wRqCYMhU6y5PK74q2sWjqSZmRkZb3MVGFXV block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   26 State  
        DB GET               26 requests for a total of     4929 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          10 /  26  (shard-cache/DB)
        SHARD CACHE         27.78% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=app.nearcrowd.near receipt_id=5csXkacKhAAQ6F6eckGqZycLgD7cY5aKHimCxw1iNfqP block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   33 State  
        DB GET               33 requests for a total of     4106 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           6 /  33  (shard-cache/DB)
        SHARD CACHE         15.38% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=app.nearcrowd.near receipt_id=3DVei5MvayiThUdrtgXDDeWRSfCG7QD859f4rtADLDNk block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   3 State  
        DB GET                3 requests for a total of      165 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          18 /   3  (shard-cache/DB)
        SHARD CACHE         85.71% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=app.nearcrowd.near receipt_id=2CZ3nibSbAFYF2rZYJdxdBc2NwLEVQyrxbfmkjsL1X14 block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   70 State  
        DB GET               70 requests for a total of   552562 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          6 requests for a total of      608 B
        STORAGE WRITE         7 requests for a total of      383 B
        TRIE NODES (guest)  235 /   2 /  69  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          15 /   1  (shard-cache/DB)
        SHARD CACHE         19.54% hit rate,  19.77% if removing 1 too large nodes from total
        CHUNK CACHE         76.80% hit rate

      process_receipt receiver=app.nearcrowd.near receipt_id=GTWiUssUkNTmHnUCHgQo91G1ML5i6y1tAWznVsoKog1z block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   26 State  
        DB GET               26 requests for a total of   544804 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          3 requests for a total of      337 B
        STORAGE WRITE         4 requests for a total of      409 B
        TRIE NODES (guest)  116 /   0 /  25  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          13 /   1  (shard-cache/DB)
        SHARD CACHE         33.33% hit rate,  34.21% if removing 1 too large nodes from total
        CHUNK CACHE         82.27% hit rate

      process_receipt receiver=27b8a4173e15829829159b39c620e9a5da908a3f68d94a70cd489188d478f37f receipt_id=GFRjGJ7J92VBZAkdSpUwN7z4v1mCmtWgrf2SU7f4MGCN block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   14 State  
        DB GET               14 requests for a total of     2519 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)          18 /  14  (shard-cache/DB)
        SHARD CACHE         56.25% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=14ca6570b456a5f904641d1005b9499aee5cfa60216bb3b73871d2256ad1cc2a receipt_id=CuNBQKsiVcZH9Nc9QcWY65oE1w4ifYtiPC5SNKDfA7fL block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   24 State  
        DB GET               24 requests for a total of     5159 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           4 /  24  (shard-cache/DB)
        SHARD CACHE         14.29% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=2cd58e2eab52381a893c42b8dc47d66633b58c16303c4398f44dbd149b6e1411 receipt_id=D3ww9qmfxoyDuJjsyGtsoCZ1MSPwhCDLtny75AsJcTMi block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   24 State  
        DB GET               24 requests for a total of     4901 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           6 /  24  (shard-cache/DB)
        SHARD CACHE         20.00% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=91505032331ede720873cf84a36609dd8755a09f816735a60ffa3ee40b4a4a67 receipt_id=G2AYqug5LN3FS2SN9GPEQERkVpK9dY6g5hQgcnSArswd block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   26 State  
        DB GET               26 requests for a total of     4795 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           6 /  26  (shard-cache/DB)
        SHARD CACHE         18.75% hit rate
        CHUNK CACHE not accessed

      process_receipt receiver=6b0021d947b74d8ab24d379f80803c5224767d3777c3d92899806dc5cecc2f77 receipt_id=88WWQXk4nrVqe2si34jS9fhouD73uaGHU5f994c46B9G block=2zB2oBqRj8dFECSaBeGP6hGGDs7KowYBVwcVfq8Ukjf9
        GET   20 State  
        DB GET               20 requests for a total of     3643 B
        DB SET                0 requests for a total of        0 B
        STORAGE READ          0 requests for a total of        0 B
        STORAGE WRITE         0 requests for a total of        0 B
        TRIE NODES (guest)    0 /   0 /   0  (chunk-cache/shard-cache/DB)
        TRIE NODES (host)           8 /  20  (shard-cache/DB)
        SHARD CACHE         28.57% hit rate
        CHUNK CACHE not accessed


